docs/DafnyRef/Types.md

```diff
@@ -1705,29 +1776,32 @@ such a `C` object can be used as a value of type `J`.
 
 ## 13.1. Type `object`
 ````grammar
-ObjectType_ = "object"
+ObjectType_ = "object" | "object?"
 ````
 
-There is a built-in trait `object` that is a supertype of all
-reference types.
+There is a built-in trait `object?` that is a supertype of all
+reference types and a trait `object` that is a supertype of all non-null reference types..
 Every class and every trait (other than `object` itself) automatically extends
-`object`. This includes types like arrays and iterators that do not permit
-explicit extending of traits. The purpose of type `object`
+`object?`; every non-null class or trait extends `object`.
+This includes types like arrays and iterators that do not permit
+explicit extending of traits. The purpose of type `object
 is to enable a uniform treatment of _dynamic frames_. In particular, it
 is useful to keep a ghost field (typically named `Repr` for
 "representation") of type `set<object>`.
 
-It serves no purpose (but does no harm) to explicitly list `object` as
+It serves no purpose (but does no harm) to explicitly list `object` ior `object?` as
 an extendee in a class or trait declaration.
 
-Trait `object` contains no members.
+Traits `object?` and  `object` contain no members.
 
 The dynamic allocation of objects is done using `new C ...`,
  where `C` is the name of a class.
  The name `C` is not allowed to be a trait,
  except that it is allowed to be `object`.
  The construction `new object` allocates a new object (of an unspecified class type).
  The construction can be used to create unique references, where no other properties of those references are needed.
+(`new object?` makes no sense; always use `new object` insteadi because the result of
+`new` is always non-null.)
```

Comment:
```suggestion
(`new object?` makes no sense; always use `new object` instead because the result of
`new` is always non-null.)
```

Well, yeah, I guess this is a fine way to think about it. I was thinking of the type following `new` follows the same kind of rule as the type follows `extends`--that is, the type (suitably parameterized) names a class-or-`object`, not a _type name_ (see comment above).

Reply:
OK